Policy crud2

aalsabag opened the pull request:
Adding max-page-size to our search calls
Added endpoint to ValidatePolicy
Updated CreatePolicy to make a call to ValidatePolicy

NEXT PR:
Add refresh setting for elasticsearch
Add filtering in ListPolicy
Add a basic, preliminary UpdatePolicy

pickjasmine reviewed (commented):
I think this looks good but you probably want someone more experienced than me to review this.

pickjasmine commented on server/server.go:
```diff
@@ -272,7 +322,11 @@ func (r *rodeServer) CreatePolicy(ctx context.Context, policyEntity *pb.PolicyEn
+	// CheckPolicy before writing to elastic
+	result, err := r.ValidatePolicy(ctx, &pb.ValidatePolicyRequest{Policy: policyEntity.RegoContent})
+	if (err != nil) || !result.Compile {
+		return nil, createError(log, "failed to compile the provided policy", err)
```
Just going to leave this comment as a reminder - can we return the same type of response that is returned from the validate policy endpoint here?

alexashley commented on proto/v1alpha1/rode.proto:
```diff
@@ -72,6 +72,12 @@ service Rode {
   }
+  rpc ValidatePolicy(ValidatePolicyRequest) returns (ValidatePolicyResponse) {
+    option (google.api.http) = {
+      post: "/v1alpha1/policies/validate"
```
I don't feel strongly about it, but for the RPC calls that don't map to restful endpoints (like attest policy or batch create) we've been using a colon to append the action, like `/v1alpha1/policies:validate"`. That seems to be the convention that Grafeas follows at least.

aalsabag replied:
@pickjasmine This change is gonna affect your PR i assume. just a heads up

alexashley commented on server/server.go:
```diff
@@ -263,6 +265,54 @@ func (r *rodeServer) ListOccurrences(ctx context.Context, occurrenceRequest *pb.
+		}, nil
+	}
+
+	// Create a new compiler instance and compile the module. Should this be moved to initialize function?
```
Making a new one might be for the best? It look like there's a lot of internal state there that might end up being shared between runs.

alexashley commented on server/server.go:
```diff
@@ -263,6 +265,54 @@ func (r *rodeServer) ListOccurrences(ctx context.Context, occurrenceRequest *pb.
+		for i := range c.Errors {
+			errorsList = append(errorsList, c.Errors[i].Error())
+		}
+		return &pb.ValidatePolicyResponse{
```
If this was a pure HTTP endpoint, I think this case and failing to parse would return a bad request (400) status, along with the errors. 

That might also remove the need for a compile flag.

aalsabag replied:
I debated on this myself for some time.
I always thought that if an endpoint behaves as it should, then it should return a 200. This is an endpoint that validates policy. It does that successfully as long as you pass in a nonempty policy.

The createpolicy endpoint is supposed to create a policy. If the policy you pass in doesn't compile, it returns a non 200 status code because it failed to do its job  i.e. create a policy.

That may be the wrong, but I wanted to get your opinion @alexashley

alexashley replied:
My only concern is that I think users might expect a 200 as meaning that the policy is valid, without bothering to look at the response body or documentation. Usually if a client provides invalid data, that would be some form of a 4xx response. 

I don't feel too strongly about it either way -- there aren't any other users of this endpoint other than the UI and it's one of those RPCs that doesn't map in a straightforward way to a REST API.

aalsabag replied:
I don't either. I'm actually quite clueless here so I'll defer to you all. I'll merge this in so we can allow jazz to merge her changes. I can change this with the update policy pr

pickjasmine replied:
I think you both make really good points. I don't feel particularly strong either way, but I agree with Alex that we might avoid some confusion switching to a 4xx status code since it's likely that there will be HTTP users of Rode.

alexashley reviewed (approved):
LGTM, just some small nits that could be in the follow up

pickjasmine commented on server/server.go:
```diff
@@ -325,7 +325,13 @@ func (r *rodeServer) CreatePolicy(ctx context.Context, policyEntity *pb.PolicyEn
 	result, err := r.ValidatePolicy(ctx, &pb.ValidatePolicyRequest{Policy: policyEntity.RegoContent})
 	if (err != nil) || !result.Compile {
-		return nil, createError(log, "failed to compile the provided policy", err)
+		message := &pb.ValidatePolicyResponse{
```
nice 💯

aalsabag commented:
Final commit. This addresses all of the comments. A couple of UI changes need to be made now.
1. A failed validation on CreatePolicy now returns 400 instead of 500
2. ValidatePolicy will now return 400 on any unsuccessful compilations.

pickjasmine commented on server/server.go:
```diff
@@ -276,14 +276,16 @@ func (r *rodeServer) ValidatePolicy(ctx context.Context, policy *pb.ValidatePoli
 			Errors:  []string{err.Error()},
-		}, nil
+		}
+		s, _ := status.New(codes.InvalidArgument, "failed to parse the policy").WithDetails(message)
```
Could this message match the others? `failed to compile the policy`

pickjasmine replied:
Actually I'm looking at this more, and I'm not sure what's appropriate here. What's the difference between a policy failing to compile and a policy failing to parse?

aalsabag replied:
I can make them match. I guess it's semantics.

Failed to parse means it failed to generate and ast. Symbols that don't exist, no matching parentheses, etc

Failed to compile means that your code is correct but you didn't declare variables, setting stings to ints...

pickjasmine replied:
Ohhh gotcha gotcha, thanks! You can keep them separate if you want, I can handle the difference in the UI just fine. I was under the impression they were the same sort of thing, but I could see value in returning two different messages for people who might be writing policy on their own.
